test/src/lib/backoff.spec.ts

```diff
@@ -0,0 +1,163 @@
+import { expect } from 'chai';
+
+import Failer from '../../lib/failer';
+
+import { StatusError } from '../../../src/lib/errors';
+import { withBackoff, OnRetryOpts } from '../../../src/lib/backoff';
+
+describe('lib/backoff', async () => {
+	const DEFAULT_OPTIONS = {
+		maxRetries: 5,
+		maxDelay: 50,
+		minDelay: 1,
+	};
+
+	it('resolves after 3 retries', async () => {
+		// Create a function that will fail 3 times so it succeeds on the 4th
+		const failer = new Failer(3);
+		// Wrap function withBackoff
+		const fnWithBackoff = withBackoff(async () => {
+			return failer.willResolve('data that fails a lot...');
+		}, DEFAULT_OPTIONS);
+		// Call the function
+		await expect(fnWithBackoff()).to.eventually.equal(
+			'data that fails a lot...',
+		);
+		// Check that the failing function was called 4 times (succeeded on 4th time)
+		expect(failer.counter).to.equal(4);
+	});
+
+	it('never waits less then minDelay', async () => {
+		const failer = new Failer(3);
+		// Track the delay responses in the retry callback
+		let minRecordedDelay: number;
+		// Setup the function
+		const fnWithBackoff = withBackoff(
+			async () => {
+				return failer.willResolve('ok');
+			},
+			{
+				...DEFAULT_OPTIONS,
+				onRetry: (data: OnRetryOpts) => {
+					if (!minRecordedDelay) {
+						minRecordedDelay = data.delay;
+					} else if (data.delay < minRecordedDelay) {
+						minRecordedDelay = data.delay;
+					}
+				},
+			},
+		);
+		// Call the function
+		await expect(fnWithBackoff()).to.eventually.equal('ok');
+		// Check that the smallest recorded delay is greater than the minDelay passed in the options
+		expect(minRecordedDelay!).to.be.greaterThan(DEFAULT_OPTIONS.minDelay);
```

Comment:
I think using Sinon fake timers is a better option here and would probably end up in simpler tests. Right now this test is checking that the delay passed to `onRetry` is the not smaller than `minDelay` but it does not test that the function actually backs off for `minDelay` ms. Same thing with the other tests

https://sinonjs.org/releases/latest/fake-timers/

Reply:
Testing that the data reported by `onRetry` is also a good test to make but is not the most important behavior.